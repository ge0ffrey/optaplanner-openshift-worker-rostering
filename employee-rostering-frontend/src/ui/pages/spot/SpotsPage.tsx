/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import {DataTable, DataTableProps} from 'ui/components/DataTable';
import MultiTypeaheadSelectInput from 'ui/components/MultiTypeaheadSelectInput'
import {spotOperations} from 'store/spot';
import Spot from 'domain/Spot';
import { AppState } from 'store/types';
import { TextInput, Text, Chip, ChipGroup } from '@patternfly/react-core';
import { connect } from 'react-redux';
import Skill from 'domain/Skill';

interface StateProps extends DataTableProps<Spot> {
  tenantId: number;
  skillList: Skill[];
}

const mapStateToProps = ({ tenantData, skillList, spotList }: AppState): StateProps => ({
  title: "Spots",
  columnTitles: ["Name", "Required Skill Set"],
  tableData: spotList.spotList,
  skillList: skillList.skillList,
  tenantId: tenantData.currentTenantId
}); 

export interface DispatchProps {
  addSpot: typeof spotOperations.addSpot;
  updateSpot: typeof spotOperations.updateSpot;
  removeSpot: typeof spotOperations.removeSpot;
}

const mapDispatchToProps: DispatchProps = {
  addSpot: spotOperations.addSpot,
  updateSpot: spotOperations.updateSpot,
  removeSpot: spotOperations.removeSpot
};

export type Props = StateProps & DispatchProps;

export class SpotsPage extends DataTable<Spot, Props> {
  constructor(props: Props) {
    super(props);
    this.addData = this.addData.bind(this);
    this.updateData = this.updateData.bind(this);
    this.removeData = this.removeData.bind(this);
  }

  createNewDataInstance(): Spot {
    return {
      tenantId: this.props.tenantId,
      name: "",
      requiredSkillSet: [] 
    };
  }

  displayDataRow(data: Spot): JSX.Element[] {
    return [
      <Text key={0}>{data.name}</Text>,
      <ChipGroup key={1}>{data.requiredSkillSet.map(skill => (
        <Chip key={skill.name} isReadOnly>
          {skill.name}
        </Chip>
      ))}
      </ChipGroup>
    ];
  }
  
  editDataRow(data: Spot): JSX.Element[] {
    return [
      <TextInput key={0}
        name="name"
        defaultValue={data.name}
        aria-label="Name"
        onChange={(value) => data.name = value}
      />,
      <MultiTypeaheadSelectInput key={1}
        emptyText={"Select required skills"}
        options={this.props.skillList}
        optionToStringMap={skill => skill.name}
        defaultValue={data.requiredSkillSet}
        onChange={selected => data.requiredSkillSet = selected}
      />
    ];
  }
  
  isValid(editedValue: Spot): boolean {
    return editedValue.name.length > 0;
  }
  
  updateData(data: Spot): void {
    this.props.updateSpot(data);
  }
  
  addData(data: Spot): void {
    this.props.addSpot(data);
  }

  removeData(data: Spot): void {
    this.props.removeSpot(data);
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(SpotsPage);
