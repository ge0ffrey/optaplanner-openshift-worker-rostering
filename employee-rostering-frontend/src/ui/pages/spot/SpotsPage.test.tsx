/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { SpotsPage, Props } from './SpotsPage';

describe('Spots page', () => {
  it('should render correctly with no spots', () => {
    const spotsPage = shallow(<SpotsPage {...noSpots} />);
    expect(toJson(spotsPage)).toMatchSnapshot();
  });

  it('should render correctly with a few spots', () => {
    const spotsPage = shallow(<SpotsPage {...twoSpots} />);
    expect(toJson(spotsPage)).toMatchSnapshot();
  });

  it('should render the viewer correctly', () => {
    const spotsPage = new SpotsPage(twoSpots);
    const spot = twoSpots.tableData[1];
    const viewer = shallow(spotsPage.renderViewer(spot));
    expect(toJson(viewer)).toMatchSnapshot();
  });

  it('should render the editor correctly', () => {
    const spotsPage = new SpotsPage(twoSpots);
    const spot = twoSpots.tableData[1];
    const editor = shallow(spotsPage.renderEditor(spot));
    expect(toJson(editor)).toMatchSnapshot();
  });

  it('should return a spot with the tenantId but no name, requiredSkillSet id or version on createNewDataInstance', () => {
    const spotsPage = new SpotsPage(twoSpots);
    const spot = spotsPage.createNewDataInstance();
    expect(spot.tenantId).toEqual(twoSpots.tenantId);
    expect(spot.name).toEqual("");
    expect(spot.requiredSkillSet).toEqual([]);
    expect(spot.id).toBeUndefined();
    expect(spot.version).toBeUndefined();
  })

  it('should call addSpot on addData', () => {
    const spotsPage = new SpotsPage(twoSpots);
    const spot = {name: "Spot", requiredSkillSet: [], tenantId: 0};
    spotsPage.addData(spot);
    expect(twoSpots.addSpot).toBeCalled();
    expect(twoSpots.addSpot).toBeCalledWith(spot);
  });

  it('should call updateSpot on updateData', () => {
    const spotsPage = new SpotsPage(twoSpots);
    const spot = {name: "Spot", requiredSkillSet: [], tenantId: 0, id: 1, version: 0};
    spotsPage.updateData(spot);
    expect(twoSpots.updateSpot).toBeCalled();
    expect(twoSpots.updateSpot).toBeCalledWith(spot);
  });

  it('should call removeSpot on removeData', () => {
    const spotsPage = new SpotsPage(twoSpots);
    const spot = {name: "Spot", requiredSkillSet: [], tenantId: 0, id: 1, version: 0};
    spotsPage.removeData(spot);
    expect(twoSpots.removeSpot).toBeCalled();
    expect(twoSpots.removeSpot).toBeCalledWith(spot);
  });

  it('should treat empty name as invalid', () => {
    const spotsPage = new SpotsPage(twoSpots);
    const components = {tenantId: 0, name: "", requiredSkillSet: []};
    const result = spotsPage.isValid(components);
    expect(result).toEqual(false);
  });

  it('should treat non-empty name as valid', () => {
    const spotsPage = new SpotsPage(twoSpots);
    const components = {tenantId: 0, name: "Spot", requiredSkillSet: []};
    const result = spotsPage.isValid(components);
    expect(result).toEqual(true);
  });
});

const noSpots: Props = {
  tenantId: 0,
  title: "Spots",
  columnTitles: ["Name"],
  tableData: [],
  skillList: [],
  addSpot: jest.fn(),
  updateSpot: jest.fn(),
  removeSpot: jest.fn()
};

const twoSpots: Props = {
  tenantId: 0,
  title: "Spots",
  columnTitles: ["Name"],
  tableData: [{
    id: 0,
    version: 0,
    tenantId: 0,
    name: "Spot 1",
    requiredSkillSet: []
  },
  {
    id: 1,
    version: 0,
    tenantId: 0,
    name: "Spot 2",
    requiredSkillSet: [{tenantId: 0, name: "Skill 1"}, {tenantId: 0, name: "Skill 2"}]
  }],
  skillList: [{tenantId: 0, name: "Skill 1"}, {tenantId: 0, name: "Skill 2"}],
  addSpot: jest.fn(),
  updateSpot: jest.fn(),
  removeSpot: jest.fn()
};