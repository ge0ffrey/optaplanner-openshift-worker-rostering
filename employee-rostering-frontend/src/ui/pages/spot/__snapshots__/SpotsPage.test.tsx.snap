// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spots page should render correctly with a few spots 1`] = `
<Card>
  <div
    style={
      Object {
        "display": "grid",
        "gridTemplateColumns": "min-content auto min-content max-content",
      }
    }
  >
    <FilterComponent
      filterListParentId="filter-list"
      filters={
        Array [
          Object {
            "getComponent": [Function],
            "name": "Name",
          },
        ]
      }
      onChange={[Function]}
    />
    <span />
    <Button
      isDisabled={false}
      onClick={[Function]}
    >
      Add
    </Button>
    <Pagination
      className=""
      dropDirection="down"
      itemCount={2}
      itemsEnd={null}
      itemsStart={null}
      onFirstClick={[Function]}
      onLastClick={[Function]}
      onNextClick={[Function]}
      onPageInput={[Function]}
      onPerPageSelect={[Function]}
      onPreviousClick={[Function]}
      onSetPage={[Function]}
      page={1}
      perPage={10}
      perPageOptions={
        Array [
          Object {
            "title": "10",
            "value": 10,
          },
          Object {
            "title": "20",
            "value": 20,
          },
          Object {
            "title": "50",
            "value": 50,
          },
          Object {
            "title": "100",
            "value": 100,
          },
        ]
      }
      titles={
        Object {
          "currPage": "Current page",
          "items": "items",
          "itemsPerPage": "Items per page",
          "optionsToggle": "Select",
          "pages": "pages",
          "paginationTitle": "Pagination",
          "perPageSuffix": "per page",
          "toFirstPage": "Go to first page",
          "toLastPage": "Go to last page",
          "toNextPage": "Go to next page",
          "toPreviousPage": "Go to previous page",
        }
      }
      toggleTemplate={[Function]}
      variant="top"
      widgetId="pagination-options-menu-top"
    />
  </div>
  <Table
    borders={true}
    caption="Spots"
    cells={
      Array [
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Name",
          "transforms": Array [
            [Function],
          ],
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "",
          "transforms": undefined,
        },
      ]
    }
    className=""
    contentId="expanded-content"
    dropdownDirection="down"
    dropdownPosition="right"
    expandId="expandable-toggle"
    gridBreakPoint="grid-md"
    onCollapse={null}
    onExpand={null}
    onSort={[Function]}
    role="grid"
    rowLabeledBy="simple-node"
    rows={
      Array [
        Object {
          "cells": Array [
            Object {
              "title": <EditableComponent
                editor={
                  <TextInput
                    aria-label="Name"
                    className=""
                    defaultValue="Spot 1"
                    isDisabled={false}
                    isReadOnly={false}
                    isRequired={false}
                    isValid={true}
                    name="name"
                    onChange={[Function]}
                    type="text"
                  />
                }
                isEditing={false}
                viewer={
                  <Text>
                    Spot 1
                  </Text>
                }
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <MultiTypeaheadSelectInput
                    defaultValue={Array []}
                    emptyText="Select required skills"
                    onChange={[Function]}
                    optionToStringMap={[Function]}
                    options={
                      Array [
                        Object {
                          "name": "Skill 1",
                          "tenantId": 0,
                        },
                        Object {
                          "name": "Skill 2",
                          "tenantId": 0,
                        },
                      ]
                    }
                  />
                }
                isEditing={false}
                viewer={
                  <ChipGroup
                    className=""
                    collapsedText="\${remaining} more"
                    expandedText="Show Less"
                    withToolbar={false}
                  />
                }
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <span>
                    <Button
                      aria-label="Save"
                      onClick={[Function]}
                      variant="link"
                    >
                      <SaveIcon
                        color="currentColor"
                        size="sm"
                        title={null}
                      />
                    </Button>
                    <Button
                      aria-label="Cancel"
                      onClick={[Function]}
                      variant="link"
                    >
                      <CloseIcon
                        color="currentColor"
                        size="sm"
                        title={null}
                      />
                    </Button>
                  </span>
                }
                isEditing={false}
                viewer={
                  <span>
                    <Button
                      aria-label="Edit"
                      onClick={[Function]}
                      variant="link"
                    >
                      <EditIcon
                        color="currentColor"
                        size="sm"
                        title={null}
                      />
                    </Button>
                    <Button
                      aria-label="Delete"
                      onClick={[Function]}
                      variant="link"
                    >
                      <TrashIcon
                        color="currentColor"
                        size="sm"
                        title={null}
                      />
                    </Button>
                  </span>
                }
              />,
            },
          ],
        },
        Object {
          "cells": Array [
            Object {
              "title": <EditableComponent
                editor={
                  <TextInput
                    aria-label="Name"
                    className=""
                    defaultValue="Spot 2"
                    isDisabled={false}
                    isReadOnly={false}
                    isRequired={false}
                    isValid={true}
                    name="name"
                    onChange={[Function]}
                    type="text"
                  />
                }
                isEditing={false}
                viewer={
                  <Text>
                    Spot 2
                  </Text>
                }
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <MultiTypeaheadSelectInput
                    defaultValue={
                      Array [
                        Object {
                          "name": "Skill 1",
                          "tenantId": 0,
                        },
                        Object {
                          "name": "Skill 2",
                          "tenantId": 0,
                        },
                      ]
                    }
                    emptyText="Select required skills"
                    onChange={[Function]}
                    optionToStringMap={[Function]}
                    options={
                      Array [
                        Object {
                          "name": "Skill 1",
                          "tenantId": 0,
                        },
                        Object {
                          "name": "Skill 2",
                          "tenantId": 0,
                        },
                      ]
                    }
                  />
                }
                isEditing={false}
                viewer={
                  <ChipGroup
                    className=""
                    collapsedText="\${remaining} more"
                    expandedText="Show Less"
                    withToolbar={false}
                  >
                    <Chip
                      className=""
                      closeBtnAriaLabel="close"
                      component="div"
                      isOverflowChip={false}
                      isReadOnly={true}
                      onClick={[Function]}
                      tooltipPosition="top"
                    >
                      Skill 1
                    </Chip>
                    <Chip
                      className=""
                      closeBtnAriaLabel="close"
                      component="div"
                      isOverflowChip={false}
                      isReadOnly={true}
                      onClick={[Function]}
                      tooltipPosition="top"
                    >
                      Skill 2
                    </Chip>
                  </ChipGroup>
                }
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <span>
                    <Button
                      aria-label="Save"
                      onClick={[Function]}
                      variant="link"
                    >
                      <SaveIcon
                        color="currentColor"
                        size="sm"
                        title={null}
                      />
                    </Button>
                    <Button
                      aria-label="Cancel"
                      onClick={[Function]}
                      variant="link"
                    >
                      <CloseIcon
                        color="currentColor"
                        size="sm"
                        title={null}
                      />
                    </Button>
                  </span>
                }
                isEditing={false}
                viewer={
                  <span>
                    <Button
                      aria-label="Edit"
                      onClick={[Function]}
                      variant="link"
                    >
                      <EditIcon
                        color="currentColor"
                        size="sm"
                        title={null}
                      />
                    </Button>
                    <Button
                      aria-label="Delete"
                      onClick={[Function]}
                      variant="link"
                    >
                      <TrashIcon
                        color="currentColor"
                        size="sm"
                        title={null}
                      />
                    </Button>
                  </span>
                }
              />,
            },
          ],
        },
      ]
    }
    sortBy={Object {}}
    variant={null}
  >
    <TableHeader
      className=""
    />
    <TableBody
      className=""
      onRowClick={[Function]}
      rowKey="id"
    />
  </Table>
</Card>
`;

exports[`Spots page should render correctly with no spots 1`] = `
<Card>
  <div
    style={
      Object {
        "display": "grid",
        "gridTemplateColumns": "min-content auto min-content max-content",
      }
    }
  >
    <FilterComponent
      filterListParentId="filter-list"
      filters={
        Array [
          Object {
            "getComponent": [Function],
            "name": "Name",
          },
        ]
      }
      onChange={[Function]}
    />
    <span />
    <Button
      isDisabled={false}
      onClick={[Function]}
    >
      Add
    </Button>
    <Pagination
      className=""
      dropDirection="down"
      itemCount={0}
      itemsEnd={null}
      itemsStart={null}
      onFirstClick={[Function]}
      onLastClick={[Function]}
      onNextClick={[Function]}
      onPageInput={[Function]}
      onPerPageSelect={[Function]}
      onPreviousClick={[Function]}
      onSetPage={[Function]}
      page={1}
      perPage={10}
      perPageOptions={
        Array [
          Object {
            "title": "10",
            "value": 10,
          },
          Object {
            "title": "20",
            "value": 20,
          },
          Object {
            "title": "50",
            "value": 50,
          },
          Object {
            "title": "100",
            "value": 100,
          },
        ]
      }
      titles={
        Object {
          "currPage": "Current page",
          "items": "items",
          "itemsPerPage": "Items per page",
          "optionsToggle": "Select",
          "pages": "pages",
          "paginationTitle": "Pagination",
          "perPageSuffix": "per page",
          "toFirstPage": "Go to first page",
          "toLastPage": "Go to last page",
          "toNextPage": "Go to next page",
          "toPreviousPage": "Go to previous page",
        }
      }
      toggleTemplate={[Function]}
      variant="top"
      widgetId="pagination-options-menu-top"
    />
  </div>
  <Table
    borders={true}
    caption="Spots"
    cells={
      Array [
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Name",
          "transforms": Array [
            [Function],
          ],
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "",
          "transforms": undefined,
        },
      ]
    }
    className=""
    contentId="expanded-content"
    dropdownDirection="down"
    dropdownPosition="right"
    expandId="expandable-toggle"
    gridBreakPoint="grid-md"
    onCollapse={null}
    onExpand={null}
    onSort={[Function]}
    role="grid"
    rowLabeledBy="simple-node"
    rows={Array []}
    sortBy={Object {}}
    variant={null}
  >
    <TableHeader
      className=""
    />
    <TableBody
      className=""
      onRowClick={[Function]}
      rowKey="id"
    />
  </Table>
</Card>
`;

exports[`Spots page should render the editor correctly 1`] = `
<tr>
  <td
    key="0"
  >
    <TextInput
      aria-label="Name"
      className=""
      defaultValue="Spot 2"
      isDisabled={false}
      isReadOnly={false}
      isRequired={false}
      isValid={true}
      key="0"
      name="name"
      onChange={[Function]}
      type="text"
    />
  </td>
  <td
    key="1"
  >
    <MultiTypeaheadSelectInput
      defaultValue={
        Array [
          Object {
            "name": "Skill 1",
            "tenantId": 0,
          },
          Object {
            "name": "Skill 2",
            "tenantId": 0,
          },
        ]
      }
      emptyText="Select required skills"
      key="1"
      onChange={[Function]}
      optionToStringMap={[Function]}
      options={
        Array [
          Object {
            "name": "Skill 1",
            "tenantId": 0,
          },
          Object {
            "name": "Skill 2",
            "tenantId": 0,
          },
        ]
      }
    />
  </td>
</tr>
`;

exports[`Spots page should render the viewer correctly 1`] = `
<tr>
  <td
    key="0"
  >
    <Text
      key="0"
    >
      Spot 2
    </Text>
  </td>
  <td
    key="1"
  >
    <ChipGroup
      className=""
      collapsedText="\${remaining} more"
      expandedText="Show Less"
      key="1"
      withToolbar={false}
    >
      <Chip
        className=""
        closeBtnAriaLabel="close"
        component="div"
        isOverflowChip={false}
        isReadOnly={true}
        key="Skill 1"
        onClick={[Function]}
        tooltipPosition="top"
      >
        Skill 1
      </Chip>
      <Chip
        className=""
        closeBtnAriaLabel="close"
        component="div"
        isOverflowChip={false}
        isReadOnly={true}
        key="Skill 2"
        onClick={[Function]}
        tooltipPosition="top"
      >
        Skill 2
      </Chip>
    </ChipGroup>
  </td>
</tr>
`;
